craft_archives/repo/apt_key_manager.py

```diff
@@ -87,53 +121,46 @@ def get_key_fingerprints(cls, *, key: str) -> List[str]:
 
         :returns: List of key fingerprints/IDs.
         """
-        # pyright: reportUnknownMemberType=false, reportUnknownVariableType=false
-        with tempfile.NamedTemporaryFile(suffix="keyring") as temp_file:
-            return (  # type: ignore
-                gnupg.GPG(keyring=temp_file.name).import_keys(key_data=key).fingerprints
-            )
+        response = _call_gpg(
+            "--show-keys", "--with-colons", stdin=key.encode()
+        ).splitlines()
+        fingerprints: List[str] = []
+        for line in response:
+            if line.startswith(b"fpr:"):
+                fingerprints.append(line[4:].decode().strip(":"))
+        return fingerprints
 
     @classmethod
     def is_key_installed(
-        cls, *, key_id: str, keyring_path: Optional[pathlib.Path] = None
+        cls, *, key_id: str, keyring_path: pathlib.Path = KEYRINGS_PATH
     ) -> bool:
         """Check if specified key_id is installed.
 
         :param key_id: Key ID to check for.
-        :param keyring_path: An optional path to the keyring to check.
-          If not provided, defaults to _DEFAULT_KEYRING.
+        :param keyring_path: An optional override to check for the keyring.
 
         :returns: True if key is installed.
         """
-        keyring_file = keyring_path or _DEFAULT_KEYRING
-
+        keyring_file = get_keyring_path(key_id, base_path=keyring_path)
         # Check if the keyring file exists first, otherwise the gpg check itself
         # creates it.
         if not keyring_file.is_file():
             logger.debug(f"Keyring file not found: {keyring_file}")
             return False
 
-        cmd = [
-            *_GPG_PREFIX,
-            "--keyring",
-            _gnupg_ring(keyring_file),
-            "--list-key",
-            key_id,
-        ]
+        # Ensure the keyring file contains the correct key
         try:
-            logger.debug(f"Executing command: {cmd}")
-            subprocess.run(
-                cmd,
-                stdout=subprocess.PIPE,
-                stderr=subprocess.STDOUT,
-                check=True,
-            )
+            logger.debug("Listing keys in keyring...")
+            _call_gpg("--list-keys", key_id, keyring=keyring_file)
         except subprocess.CalledProcessError as error:
             # From testing, the gpg call will fail with return code 2 if the key
             # doesn't exist in the keyring, so it's an expected possible case.
             _expected_returncode = 2
             if error.returncode != _expected_returncode:
-                logger.warning(f"Unexpected gpg failure: {error.output}")
+                logger.warning(f"Unexpected gpg failure: {error.output.decode()}")
+            logger.warning(
```

Comment:
The first time a new repo is hit, the keyring file won't exist so we'll never call gpg on it. This only occurs when the keyring file exists but doesn't contain the right key. This is an error, but since we know how to fix it (overwrite the file), I feel warning is sufficient.

That said... thinking through this further, I think I might need to change the handling here, lest `install_key` fail. As the code is currently written, many possible return codes from GPG (including code 2 if gpg finds the file but it's not a valid keyring) will spit out a warning here and then fail during the installation step. Roughly approximating the behaviour of a key installation with a corrupted keyring file:

```bash
lengau@hyperion:/tmp$ echo 'This is invalid data!' > /tmp/not_a_real.gpg
lengau@hyperion:/tmp$ gpg --batch --no-default-keyring --show-keys /tmp/not_a_real.gpg 
gpg: no valid OpenPGP data found.
lengau@hyperion:/tmp$ echo $?
2
lengau@hyperion:/tmp$ gpg --batch --no-default-keyring --keyring /tmp/not_a_real.gpg --import ~/Work/Code/craft-archives/tests/integration/repo/test_keys/FC42E99D.asc 
gpg: [don't know]: invalid packet (ctb=54)
gpg: keydb_get_keyblock failed: Value not found
gpg: [don't know]: invalid packet (ctb=54)
gpg: /tmp/not_a_real.gpg: copy to '/tmp/not_a_real.gpg.tmp' failed: Invalid packet
gpg: error writing keyring '/tmp/not_a_real.gpg': Invalid packet
gpg: error reading '/home/lengau/Work/Code/craft-archives/tests/integration/repo/test_keys/FC42E99D.asc': Invalid packet
gpg: import from '/home/lengau/Work/Code/craft-archives/tests/integration/repo/test_keys/FC42E99D.asc' failed: Invalid packet
gpg: Total number processed: 0

lengau@hyperion:/tmp$ echo $?
2
```

In so doing though, I'd like to consider options for more robust handling of gpg commands. The `gnupg` library we were using before (for admittedly a single call) is both an old version and unmaintained, so perhaps `python-gnupg` is the best option? Or we could use the python interface to `gpgme`. Or maybe just improving `_call_gpg` using [their documentation](https://github.com/gpg/gnupg/blob/master/doc/DETAILS) (notably by using `--with-colons` everywhere and the `status-fd` output).

Perhaps for now I should just turn the `CalledProcessError` into an `AptGPGKeyringError` that says the file's corrupt and includes gpg's stderr?

Reply:
I decided the best way to handle it right now is to eliminate the distinction in the return codes and just make the key installer remove the file if it exists but doesn't contain the key. Since each keyring should contain only one key this shouldn't make a practical difference.

Reply:
> The first time a new repo is hit, the keyring file won't exist so we'll never call gpg on it.

You are right of course! Slipped my mind.
I agree with the rest of your reasoning, thanks!